llvm/lib/CAS/MappedFileRegionBumpPtr.cpp

```diff
@@ -146,66 +157,79 @@ Expected<MappedFileRegionBumpPtr> MappedFileRegionBumpPtr::create(
   if (!FileSize)
     return createFileError(Result.Path, FileSize.getError());
 
+  // If the size is smaller than the capacity, we need to initialize the file.
+  // It maybe empty, or may have been shrunk during a previous close.
   if (FileSize->Size < Capacity) {
     // Lock the file exclusively so only one process will do the initialization.
-    if (Error E = InitLock.unlock())
-      return std::move(E);
     if (Error E = InitLock.lock(sys::fs::LockKind::Exclusive))
       return std::move(E);
     // Retrieve the current size now that we have exclusive access.
     FileSize = FileSizeInfo::get(File);
     if (!FileSize)
-        return createFileError(Result.Path, FileSize.getError());
+      return createFileError(Result.Path, FileSize.getError());
   }
 
-  // At this point either the file is still under-sized, or we have the size for
-  // the completely initialized file.
-
-  if (FileSize->Size < Capacity) {
-    // We are initializing the file; it may be empty, or may have been shrunk
-    // during a previous close.
-    // FIXME: Detect a case where someone opened it with a smaller capacity.
+  uint64_t MappingSize = FileSize->Size;
+  // If the size is still smaller than the minimal required size, we need to
+  // resize the file to the capacity.
+  if (FileSize->Size < MinCapacity) {
     assert(InitLock.Locked == sys::fs::LockKind::Exclusive);
```

Comment:
Why was the comment **before** the FIXME removed? It still seems relevant here.

Reply:
File won't be shrunk to smaller than MinSize. This is the code path to initialize new file.